llvm/lib/Transforms/Yk/ModuleClone.cpp

```diff
@@ -58,53 +53,77 @@ struct YkModuleClone : public ModulePass {
   YkModuleClone() : ModulePass(ID) {
     initializeYkModuleClonePass(*PassRegistry::getPassRegistry());
   }
-  void renameFunctions(Module &ClonedModule) {
-    for (llvm::Function &F : ClonedModule) {
+
+  /**
+   * @brief Clones eligible functions within the given module.
+   *
+   * This function iterates over all functions in the provided LLVM module `M`
+   * and clones those that meet the following criteria:
+   *
+   * - The function does **not** have external linkage and is **not** a
+   * declaration.
+   * - The function's address is **not** taken.
+   *
+   * @param M The LLVM module containing the functions to be cloned.
+   * @return A map where the keys are the original function names and the
+   *         values are pointers to the cloned `Function` objects. Returns
+   *         a map if cloning succeeds, or `nullopt` if cloning fails.
+   */
+  std::optional<std::map<std::string, Function *>>
+  cloneFunctionsInModule(Module &M) {
+    LLVMContext &Context = M.getContext();
+    std::map<std::string, Function *> ClonedFuncs;
+    for (llvm::Function &F : M) {
+      // Skip external declarations.
       if (F.hasExternalLinkage() && F.isDeclaration()) {
         continue;
       }
-      // Skip functions that are address taken
-      if (!F.hasAddressTaken()) {
-        F.setName(Twine(YK_CLONE_PREFIX) + F.getName());
+      // Skip already cloned functions or functions with address taken.
+      if (F.hasAddressTaken() || F.getName().startswith(YK_CLONE_PREFIX)) {
+        continue;
+      }
+      ValueToValueMapTy VMap;
+      Function *ClonedFunc = CloneFunction(&F, VMap);
+      if (ClonedFunc == nullptr) {
+        Context.emitError("Failed to clone function: " + F.getName());
+        return std::nullopt;
       }
+      // Copy arguments
+      auto DestArgIt = ClonedFunc->arg_begin();
+      for (const Argument &OrigArg : F.args()) {
+        DestArgIt->setName(OrigArg.getName());
+        VMap[&OrigArg] = &*DestArgIt++;
+      }
```

Comment:
Is this needed? From what I can see `llvm::CloneFunction` already copies the arguments. And the `VMap` isn't actually read anywhere here.

Reply:
ok will remove.

Reply:
removed :point_right: 4563f62f6b46eca4d91cd71192c4a32356690560

Reply:
I don't see the commit.